### ISSUE TITLE
Question about impl of ABC

### ISSUE THREAD

**Issue by yzhangcs** [Shark-NLP/CAB#1]
@z562 Hi, thank you for your great work.
I have some questions about the casual impl of ABC
https://github.com/Shark-NLP/CAB/blob/main/efficient-attention/efficient_attention/modules/abc.py#L119-L121
which seems to be not strictly aligned with the descriptions in the paper?
<img width="644" alt="image" src="https://github.com/Shark-NLP/CAB/assets/18402347/a915cac9-f45b-4cec-a81a-6374e6fbc62a">
According to appendix B.1, my understanding is that, at time step $t$ we should sum up all $\phi_i\otimes k_i$ for $i \leq t$, and then do normalization with $z_t = \sum_i^t \phi_i$. In contrast, your impl normalizes the intermediate results at each step with $z_i$, and is not strictly in consistant with the paper.

**Comment by z562** [comment]
At step $t$, ABC sum up all $\phi_i \otimes k_i, i \leq t$ and do normalization with $z^{t} = \sum^t \phi_i$. At step $t-1$, They will be $\phi_i \otimes k_i, i \leq t-1$ and $z_t = \sum^{t-1} \phi_i$. As a causal attention, ABC needs to do normalization at each step $t$.

**Comment by yzhangcs** [comment]
@z562 Hi, what I mean is that in your impl, at step $t$, the final result is $\sum_i^t\phi_i \otimes k_i / z_i$, but I think we should do summation first and then divide it by the denominator at $t$, i.e. $(\sum_i^t\phi_i \otimes k_i) / z_t$?

**Comment by z562** [comment]
https://github.com/Shark-NLP/CAB/blob/main/efficient-attention/efficient_attention/modules/abc.py#L119
At the step $t$, L120 calculates the $\text{alpha\\_k} = \phi_i / z_t, i \leq t$. At the step t, L120 and L121 are to obtain $\text{k\\_logits}=(\phi_i / z_t) \otimes k_t = (\phi_i \otimes k_t) / z_t$.  Changing the calculation order $z_t$ results in the equivalent attention score.

**Comment by yzhangcs** [comment]
Got what you mean, thank you
